cpp/src/join/joining.cu

```diff
@@ -253,260 +220,315 @@ gdf_error join_call( int num_cols, gdf_column **leftcol, gdf_column **rightcol,
-  bool const return_output_indices{(nullptr != left_indices) and
-                                   (nullptr != right_indices)};
+  std::vector<gdf_size_type> right_columns_in_common (columns_in_common.size());
 
-  GDF_REQUIRE(construct_output_dataframe or return_output_indices,
-              GDF_INVALID_API_CALL);
+  for (unsigned int i = 0; i < columns_in_common.size(); ++i)
+  {
+      right_columns_in_common [i] = columns_in_common[i].second;
+  }
 
-  auto const left_col_size = left_cols[0]->size;
-  auto const right_col_size = right_cols[0]->size;
+  cudf::table empty_left = cudf::empty_like(left);
+  cudf::table empty_right = cudf::empty_like(right);
+  std::vector <gdf_size_type> right_non_common_indices = non_common_column_indices(right.num_columns(),
+                                                                         right_columns_in_common);;
+  cudf::table tmp_right_table = empty_right.select(right_non_common_indices);
+  cudf::table tmp_table = cudf::concat(empty_left, tmp_right_table);
+  
+  // If there is nothing to join, then send empty table with all columns
+  if (left_on.empty() || right_on.empty() || left_on.size() != right_on.size())
+  {
+      return tmp_table;
+  }
 
+  // Even though the resulting table might be empty, but the column should match the expected dtypes and other necessary information
+  // So, there is a possibility that there will be lesser number of right columns, so the tmp_table.
   // If the inputs are empty, immediately return
-  if ((0 == left_col_size) && (0 == right_col_size)) {
-    return GDF_SUCCESS;
+  if ((0 == left.num_rows()) && (0 == right.num_rows())) {
+      return tmp_table;
   }
 
   // If left join and the left table is empty, return immediately
-  if ((JoinType::LEFT_JOIN == join_type) && (0 == left_col_size)) {
-    return GDF_SUCCESS;
+  if ((JoinType::LEFT_JOIN == join_type) && (0 == left.num_rows())) {
+      return tmp_table;
   }
 
   // If Inner Join and either table is empty, return immediately
   if ((JoinType::INNER_JOIN == join_type) &&
-      ((0 == left_col_size) || (0 == right_col_size))) {
-    return GDF_SUCCESS;
+      ((0 == left.num_rows()) || (0 == right.num_rows()))) {
+      return cudf::empty_like(tmp_table);
   }
 
-
   //if the inputs are nvcategory we need to make the dictionaries comparable
-  bool at_least_one_category_column = false;
-  for(int join_column_index = 0; join_column_index < num_cols_to_join; join_column_index++){
-    at_least_one_category_column |= left_cols[left_join_cols[join_column_index]]->dtype == GDF_STRING_CATEGORY;
-  }
+  bool at_least_one_category_column = std::any_of (cbegin(left_on), cend(left_on),
+                                                  [&](auto index) {  return (left.get_column(index)->dtype == GDF_STRING_CATEGORY);});
+  
+  std::vector<gdf_column*> new_left_cols(left.num_columns());
+  std::vector<gdf_column*> new_right_cols(right.num_columns());
```

Comment:
```suggestion
  std::vector<gdf_column*> new_right_cols(right.begin(), right.end());
```

Reply:
as `left` and `right` are `const`s, we can't pass begin and end to update the vector which is just `gdf_column*`